### PR TITLE
Optimise dockerfiles, fix ci, release images

### LINKED ISSUE #439: Optimise dockerfiles (multi-stage, caching)
- The dockerfiles could be optimised to be multi-stage & grouping commands into layers.
- We should also create layers that can be cached easily first, then add layers that break cache later in the file.
- While doing this, there is no harm adding more descriptive labels to images too.

I'm happy to contribute this if you assign me 😄 

Also, I am getting errors trying to build the underpass image (I can also look into this when optimising, if needed):

```
8.207 checking for ld... /usr/bin/ld -m elf_x86_64
8.207 checking if the linker (/usr/bin/ld -m elf_x86_64) is GNU ld... yes
8.208 checking for shared library run path origin... done
8.222 checking 32-bit host C ABI... no
8.231 checking how to run the C preprocessor... gcc -E
8.260 checking for ELF binary format... yes
8.265 checking for the common suffixes of directories in the library search path... lib,lib,lib64
8.277 checking for CFPreferencesCopyAppValue... no
8.288 checking for CFLocaleCopyPreferredLanguages... no
8.299 checking for GNU gettext in libc... yes
8.338 checking whether to use NLS... yes
8.338 checking where the gettext function comes from... libc
8.353 checking that generated files are newer than configure... done
8.353 configure: creating ./config.status
8.691 config.status: error: cannot find input file: `Makefile.in'
------
failed to solve: process "/bin/sh -c ../configure &&     make -j $(nproc) &&     make install" did not complete successfully: exit code: 1
```

### PR DESCRIPTION
**Pull request by spwoodcock**
Fixes #439

- Updated to multi-stage builds.
- Only installed required **runtime** libs for final underpass image.
- Reordered layers for more efficient caching.
- Added labels.

Significant image size reductions:
- api 1.31GB --> 228MB
- underpass 2.5GB --> 1.88GB
- ui 1.1GB --> 18.5MB

The underpass image could probably be reduced further by selective usage of boost libraries, instead of `libboost-all-dev`.

### Note

With the images being named ghcr.io/xxx/xxx it would be possible to push them to the underpass repo container registry.

## Update 09/11/2023

- Added stages to image:
  - `debug` for local testing.
  - `ci` for ci tests.
  - `prod` for production image. 
- Added workflow for publishing release images.
- Set ci_local.sh to use the `ci` image, with bundled entrypoint.
- Added github workflow local testing via [act](https://github.com/nektos/act)
- Made ports user configurable via env vars.
- Made image names configurable via env vars.
- Note: ci_local uses docker compose V2 (`docker compose` instead of `docker-compose`)
  - It's the standard tool now bundled with docker install (a re-implementation in Go instead of Python, better in every way).

### CODE REVIEW

**Review by emi420 — COMMENTED** [review]
The build is working, but some things are failing or missing. 

-  The directory `/code` changed, and with this, the bootstrap script stopped working for Docker
- The API port changed, so the UI is not working now
- The source files and the utils folder is missing, so the bootstrap script is not available anymore

Keep in mind that the Docker image is not only for deployment, but also for running Underpass locally. Maybe we can have two different Docker images, one for deployment and one for development/testing/running locally.

File: docker-compose.yml [hunk 1]
```diff
@@ -74,14 +76,15 @@ services:
...
+    image: "ghcr.io/hotosm/underpass/ui:${APP_VERSION:-0.1.0}"
+    container_name: underpass_ui
     build:
      context: .
      dockerfile: docker/underpass-ui.dockerfile
     # volumes:
     #   - js:/code
     ports:
-      - "5000:5000"
+      - "8080:8080"
```
[the remark this answers is not in the document]
Comment (spwoodcock) [reply docker-compose.yml:87]: I can use 5000:8080 if you prefer?

The underlying tool used to serve changed.
Before you we bundling all node_modules, then runing `yarn cosmos` for the dev server.
The new dockerfile builds the dist, then serves via a webserver, which runs on port 8080.

File: docker/underpass-ui.dockerfile [hunk 2]
```diff
@@ -1,15 +1,66 @@
...
+#
+ARG NODE_TAG=18
 
-LABEL maintainer="Humanitarian OpenStreetMap Team" Description="This image provides the Underpass UI playground" Vendor="HOT" Version="dev"
 
-RUN apk --no-cache add git
+FROM docker.io/node:${NODE_TAG}-bookworm-slim as base
+ARG APP_VERSION
+ARG COMMIT_REF
+LABEL org.hotosm.fmtm.app-name="underpass-ui" \
```
[the remark this answers is not in the document]
Comment (spwoodcock) [reply docker/underpass-ui.dockerfile:23]: Thanks for the catch!

File: docker/underpass.dockerfile [hunk 3]
```diff
@@ -50,4 +87,27 @@ RUN ../configure && \
...
+    -y --no-install-recommends \
+        "libboost-all-dev" \
+        "libgdal32" \
+        "libxml++2.6-2v5" \
+        "libjemalloc2" \
+        "libpqxx-6.4" \
+        "libgumbo1" \
+    && rm -rf /var/lib/apt/lists/*
+COPY --from=build /code/build /app
+WORKDIR /app
```
[the remark this answers is not in the document]
Comment (spwoodcock) [reply docker/underpass.dockerfile:108]: I prefer /opt as it's more standard linux, but I updated to /code to avoid changing the CI scripts.

**Comment by spwoodcock** [comment]
The standard way to have different environments is to add different stages to the dockerfile.

I usually have a `debug` stage, and a `prod` stage. The target can be specified when you build (using docker compose, gh-workflows, etc).

**Comment by spwoodcock** [comment]
Made updates: see in the first post.

Let me know if anything else needs updating 👍

**Comment by spwoodcock** [comment]
Made the final update just now:
- Create `debug` images for local debug stage and `prod` stage for the actual releases.
- Bind mount local code into the containers so you get live updates for both fastapi and cosmos.

I can build and push the images to the hotosm container registry.
The only thing left to do @emi420 is to associate the packages with the repo (only the repo owner can do that):
https://github.com/orgs/hotosm/packages/container/package/underpass
https://github.com/orgs/hotosm/packages/container/package/underpass%2Fapi
https://github.com/orgs/hotosm/packages/container/package/underpass%2Fui
'Link this package to a repository'

**Comment by spwoodcock** [comment]
I will rebase with `master` to solve conflicts.

**Comment by spwoodcock** [comment]
Rebased - should be working.

I can't work out why the test is failing with a 403 unauthorised, when the permissions are set for the repo actions to write to the container/package repository.

I feel like it is because I am running the workflow & I don't have any member permissions on the repo.

File: docker-compose.yml [hunk 4]
```diff
@@ -58,30 +62,41 @@ services:
...
+     args:
+        APP_VERSION: ${APP_VERSION:-debug}
+    # # Mount underpass-ui repo
     # volumes:
-    #   - js:/code
+    #   - ../underpass-ui/src:/code/src
+    #   - ../underpass-ui/playground:/code/playground
     ports:
-      - "5000:5000"
+      - "${UI_PORT:-5000}:8080"
```
Comment (emi420) [inline docker-compose.yml:99]: Hey @spwoodcock , I've tested this PR and the UI is working only when I change this to  ```- "${UI_PORT:-5000}:5000"```

**Comment by emi420** [comment]
It's mostly working! with `docker-compose up -d` the UI is not working but if I change the port is working perfectly:

```
ports:
      - "${UI_PORT:-5000}:5000"
```

CI is failing, possibly because the workflow version and GitHub settings, you know better what changes are needed there.

Thanks @spwoodcock

**Review by emi420 — CHANGES_REQUESTED** [review]
It's mostly working! with docker-compose up -d the UI is not working but if I change the port is working perfectly:

ports:
      - "${UI_PORT:-5000}:5000"

CI is failing, possibly because the workflow version and GitHub settings, you know better what changes are needed there.

Thanks @spwoodcock

**Comment by spwoodcock** [comment]
That seemed to work 🎉

**Pull request merged**